src/FwdState.cc

```diff
@@ -1314,7 +1319,11 @@ FwdState::reforward()
         return 0;
     }
 
-    assert(e->store_status == STORE_PENDING);
+    if (e->store_status == STORE_OK) {
+        debugs(17, 3, "No, the entry is STORE_OK already");
+        return 0;
+    }
```

Comment:
I checked that FwdState::completed() would hit the 'e->store_status == STORE_OK' assertion (added in e322b47) and removed that assertion. In that test (where server responds with an HTTP 502 reply) I adjusted the code so that reforward() would return 'true' when called the first time (so that markStoredReplyAsWhole() does not complete the entry). Also note that this scenario should never happen in FTP because reforward() always returns false (FTP code never sets ENTRY_FWD_HDR_WAIT).